src/backend/common/dispatch.hpp

```diff
@@ -42,3 +44,73 @@ inline T greatestPrimeFactor(T n) {
 
     return v;
 }
+
+// For OPENCL, the dimensions of local are returned
+// usage: cl::NDRange local = bestBlockSize<cl::NDRange>(dims, WG)
+// For CUDA, the dimensions of 1 block are returned
+// usage: dim3 block = bestBlockSize<dim3>(dims, 32);
+// The parameter dims can have any type as long as it is convertable to unsigned
+
+// Remark: The bestBlockSize is only best for independent element operations, as
+// are: copying, scaling, math on independent elements, ...
+// Since vector dimensions can be returned, it is NOT USABLE FOR BLOCK
+// OPERATIONS, as are: matmul, etc.
+template<typename Tout, typename Tin>
+Tout bestBlockSize(const Tin dims[4], unsigned warp) {
+    const unsigned d0         = static_cast<unsigned>(dims[0]);
+    const unsigned d1         = static_cast<unsigned>(dims[1]);
+    const unsigned d2         = static_cast<unsigned>(dims[2]);
+    const unsigned OCC        = 3;
+    const unsigned elements   = d0 * d1;
+    const unsigned minThreads = warp / 4;  // quarter wave
```

Comment:
NVIDIA & AMD GPU's are build on top of processors using SIMD.  For all GPU's of NVIDIA 1 instruction & 8 datasets, for AMD 1 instruction & 16 datasets.
I continue for NVIDIA, although the same for AMD.
- The processor now executes as follows:
   Wave 1
   - 1st instruction / 0...7 datasets (= 8 threads = quarter wave)
   - 1st instruction / 8...15 datasets (= 8 threads = quarter wave)
   - 1st instruction / 16...23 datasets (= 8 threads = quarter wave)
   - 1st instruction / 24...31 datasets (= 8 threads = quarter wave)
   Wave 2
   - 2nd instruction / 0...7 datasets (= 8 threads)
   - ...
   - 
- By repeating the same instruction 4x, we get the impression that a full warp is executed in parallel.
- Suppose only thread 24 exits, than the processor will service datasets 25...31 each with an 1instruction and 1 dataset  --> 7 instructions for 7 datasets, iso 1 instruction for 8 datasets.
- Suppose datasets 24...31 exit together, than everything will continue although now only with a repetition of 3x.  Each instruction (SIMD) remains servicing 8 datasets.  Losing quarter waves will keep the processor at fully efficiency, on condition that caching of global memory still can follow.  From lecture, I discovered that the caching system can keep up, up to 50% of active datasets (16 in case of NVIDIA).  With lower datasets (threads) active per processor, the processor will have to wait.

Since both architectures decided to split their respective waves into 4 instructions, we can use the same formula to determine the minThreads for both NVIDIA and for AMD.


The cache line size is 128Bytes for NVIDIA and 64Bytes for AMD (A10).
The 128Bytes = 32 datasets (full warp) * 4 bytes (size float), this means that the GPU will read a full warp (linear) at once.  This is only useful, when we can read this data sequentially.  On top when 1 cache line is read, the next 3 cache lines are already ordered in the hope that we will continue reading.
Resulting from the larger cache line for NVIDIA, it will be faster in sequential reading although will be slower with random access (reading multiple buffers in parallel).  The cache lines are also use to write to global memory, although this is fully asynchronous so no waiting times (even for random access) as long as the total cache is not occupied.


How much of the above do you want into the comments?

Reply:
Based on what you explained about the constant, cache info is irrelevant for this comment.

I think something like the below should be good enough
```cpp
<return> bestBlockSize(<params>) {
    // Currently both NVIDIA and AMD hardware splits each warp/wavefront into
    // four groups of 8/16 threads, respectively.
    // NVIDIA - 32(warp) / 8       = 4
    // AMD    - 64(wavefront) / 16 = 4
    constexpr unsigned WAVE_COUNT = 4;
    const unsigned minThreads = warp / WAVE_COUNT;
    ....
}
```

Reply:
My bad, I wasn't sure if such warp sub division happens. I don't think that's how CUDA warps work, at least from what I know. @umar456 Am I missing something from what @willyborn is referring to here ?

Reply:
Updated as proposed.
Add references: 
https://www.cvg.ethz.ch/teaching/2011spring/gpgpu/GPU-Optimization.pdf
https://en.wikipedia.org/wiki/Graphics_Core_Next#SIMD_Vector_Unit

Reply:
I don't think NVIDIA has split warps since the Kepler architecture(2012). NVIDIA did this on the Fermi and earlier cards. AMD still does this to some extent for their older generations. The new RDNA architecture moved to 32 thread wavefronts and each thread is executed in parallel.

https://developer.amd.com/wp-content/resources/RDNA_Shader_ISA.pdf 

> Suppose datasets 24...31 exit together, than everything will continue although now only with a repetition of 3x. Each instruction (SIMD) remains servicing 8 datasets. Losing quarter waves will keep the processor at fully efficiency, on condition that caching of global memory still can follow. From lecture, I discovered that the caching system can keep up, up to 50% of active datasets (16 in case of NVIDIA). With lower datasets (threads) active per processor, the processor will have to wait.

Ignoring the 4 clock per wavefront point, The inactive threads in a wavefront will still execute a NOP instruction. This is even the case if only one thread is active.  See page 41-42 here: http://developer.amd.com/wordpress/media/2013/07/AMD_Accelerated_Parallel_Processing_OpenCL_Programming_Guide-rev-2.7.pdf

Reply:
Yes, I only read about such division from the days of Fermi.

Reply:
I am thinking through this again and I think I am not entirely correct here. Yes, for floating point operations, typically the entire warp is executing the same instruction except for AMD, but for instructions that are more expensive(like doubles and special functions) this sort of behavior will occur. I think @willyborn's change will help for those types of operations.